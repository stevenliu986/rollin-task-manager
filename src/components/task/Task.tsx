import React, { useState } from 'react';
import { Divider, Card, Checkbox, Button, Flex, Modal, Spin } from 'antd';
import TaskHeader from './TaskHeader';
import PrimaryButton from '../button/PrimaryButton';
import TaskDetail from './TaskDetail';
import deleteIcon from '../../assets/images/Delete Icon.png';
import type { ITask } from './AddTask';
import './task.css';
import type { CheckboxChangeEvent } from 'antd/es/checkbox';

interface ITaskProps {
    taskDetail: Readonly<ITask>;
    onDelete?: () => Promise<void> | void;
    onToggle?: (checked: boolean) => Promise<void> | void;
}

const CARD_STYLE: React.CSSProperties = {
    width: '100%',
    maxWidth: 500,
    margin: '16px auto',
    backgroundColor: '#252525',
    position: 'relative',
    borderRadius: '16px'
};

const LOADER_STYLE: React.CSSProperties = {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'rgba(0,0,0,0.5)',
    zIndex: 1
};

const DIVIDER_STYLE: React.CSSProperties = {
    margin: '12px 0',
    backgroundColor: '#767676',
    height: '2px'
};

const Task: React.FC<ITaskProps> = ({
    taskDetail,
    onDelete,
    onToggle
}) => {
    const [taskStatus, setTaskStatus] = useState<boolean>(taskDetail.complete);
    const [isDeleteModalOpen, setIsDeleteModalOpen] = useState<boolean>(false);
    const [loading, setLoading] = useState<boolean>(false);

    const handleCheckBoxChange = async (evt: CheckboxChangeEvent) => {
        const newStatus = evt.target.checked;
        setTaskStatus(newStatus);
        try {
            setLoading(true);
            onToggle?.(newStatus);
        } finally {
            setLoading(false);
        }
    };

    return (
        <Card style={CARD_STYLE}>
            {loading && (
                <div style={LOADER_STYLE}>
                    <Spin size="large" />
                </div>
            )}

            <TaskHeader taskDetail={{ ...taskDetail, complete: taskStatus }} />
            <Divider style={DIVIDER_STYLE} />

            <Flex style={{ alignItems: 'flex-start' }}>
                <Checkbox
                    className="circle-checkbox"
                    checked={taskStatus}
                    onChange={handleCheckBoxChange}
                    disabled={loading}
                    style={{
                        marginRight: '12px',
                    }}
                    aria-label={taskStatus ? "Mark task as incomplete" : "Mark task as complete"}
                />
                <TaskDetail taskDetail={taskDetail} />
                <Button
                    type="text"
                    onClick={() => setIsDeleteModalOpen(true)}
                    disabled={loading}
                    aria-label="Delete task"
                >
                    <img src={deleteIcon} alt="delete icon" />
                </Button>
            </Flex>

            <Modal
                title={<span style={{ fontSize: '1.25rem' }}>Are you sure you want to delete this task?</span>}
                open={isDeleteModalOpen}
                onCancel={() => setIsDeleteModalOpen(false)}
                styles={{
                    header: {
                        textAlign: 'center',
                        fontWeight: 500,
                        padding: '8px',
                    },
                    body: {
                        textAlign: 'center',
                        padding: '8px 24px 16px'
                    },
                    content: {
                        borderRadius: '18px'
                    }
                }}
                footer={
                    <div style={{
                        display: 'flex',
                        justifyContent: 'center',
                        gap: '1.5rem'
                    }}>
                        <PrimaryButton
                            key="cancel"
                            onClick={() => setIsDeleteModalOpen(false)}
                            style={{
                                backgroundColor: 'transparent',
                                border: '1px solid #252525',
                                color: '#252525',
                                width: '156px',
                                height: '40px',
                                fontSize: '0.875rem',
                                margin: 0
                            }}
                            aria-label="Cancel task deletion"
                        >
                            Go back
                        </PrimaryButton>
                        <PrimaryButton
                            key="delete"
                            onClick={() => {
                                setIsDeleteModalOpen(false);
                                onDelete?.();
                            }}
                            style={{
                                backgroundColor: '#FF50BE',
                                width: '156px',
                                height: '40px',
                                fontSize: '0.875rem',
                                margin: 0
                            }}
                            aria-label="Confirm task deletion"
                        >
                            Yes, delete
                        </PrimaryButton>
                    </div>
                }
                closable={false}
            >
                <p style={{ margin: 0, fontSize: '0.875rem' }}>This action cannot be undone.</p>
            </Modal>

            <div style={{
                display: 'flex',
                justifyContent: 'center',
                width: '100%',
                marginTop: '16px'
            }}>
                <PrimaryButton
                    style={{
                        height: '40px',
                        width: '263px',
                        backgroundColor: '#FFF',
                        fontSize: '1rem',
                    }}
                    disabled={loading}
                    aria-label="Manage task details"
                >
                    Manage task
                </PrimaryButton>
            </div>
        </Card>
    );
};

export default React.memo(Task);