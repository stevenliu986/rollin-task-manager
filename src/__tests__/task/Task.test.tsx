import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Task from '../../components/task/Task';
import type { ITask } from '../../components/task/AddTask';

vi.mock('./TaskHeader', () => ({
    default: ({ taskDetail }: { taskDetail: ITask }) => (
        <div data-testid="task-header">{taskDetail.title}</div>
    )
}));

vi.mock('./TaskDetail', () => ({
    default: ({ taskDetail }: { taskDetail: ITask }) => (
        <div data-testid="task-detail">{taskDetail.description}</div>
    )
}));

vi.mock('../button/PrimaryButton', () => ({
    default: ({ children, ...props }: any) => (
        <button {...props}>{children}</button>
    )
}));

vi.mock('antd', async (importOriginal) => {
    const mod = await importOriginal();
    return {
        ...mod,
        Checkbox: ({ onChange, ...props }: any) => (
            <input
                type="checkbox"
                onChange={(e) => onChange?.({ target: { checked: e.target.checked } })}
                {...props}
            />
        ),
        Modal: ({ children, ...props }: any) => (
            props.open ? <div data-testid="antd-modal">{children}</div> : null
        )
    };
});

describe('Task Component', () => {
    const mockTask: ITask = {
        id: 1,
        title: 'Test Task',
        description: 'Test Description',
        complete: false,
        created: '2023-01-01',
        due: '2023-12-31'
    };

    const mockOnToggle = vi.fn();
    const mockOnDelete = vi.fn();

    beforeEach(() => {
        mockOnToggle.mockClear();
        mockOnDelete.mockClear();
    });

    it('toggles task status when checkbox is clicked', async () => {
        render(
            <Task
                taskDetail={mockTask}
                onToggle={mockOnToggle}
                onDelete={mockOnDelete}
            />
        );

        const checkbox = screen.getByRole('checkbox');
        fireEvent.click(checkbox);

        await waitFor(() => {
            expect(mockOnToggle).toHaveBeenCalledWith(true);
        });
    });

    it('matches snapshot', () => {
        const { asFragment } = render(
            <Task
                taskDetail={mockTask}
                onToggle={mockOnToggle}
                onDelete={mockOnDelete}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});