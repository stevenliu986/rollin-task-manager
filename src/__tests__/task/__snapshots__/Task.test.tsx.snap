// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Task Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ant-card ant-card-bordered css-dev-only-do-not-override-vrrzze"
    style="width: 100%; max-width: 500px; margin: 16px auto; background-color: rgb(37, 37, 37); position: relative; border-radius: 16px;"
  >
    <div
      class="ant-card-body"
    >
      <div
        style="display: flex; justify-content: space-between; margin-bottom: 16px;"
      >
        <button
          class="ant-btn css-dev-only-do-not-override-vrrzze ant-btn-default ant-btn-color-default ant-btn-variant-outlined ant-btn-sm"
          style="background-color: rgb(25, 107, 229); border: 1px solid rgb(25, 107, 229); border-radius: 25px; padding: 0px 8px; display: flex; align-items: center;"
          type="button"
        >
          <span
            style="font-size: 0.875rem; color: rgb(255, 255, 255);"
          >
            In Progress
          </span>
        </button>
        <button
          class="ant-btn css-dev-only-do-not-override-vrrzze ant-btn-default ant-btn-color-default ant-btn-variant-outlined ant-btn-sm"
          style="background-color: rgb(245, 245, 245); border-radius: 25px; padding: 0px 8px; display: flex; align-items: center;"
          type="button"
        >
          <img
            alt="leading-icon"
            src="/src/assets/images/Leading%20Icon.png"
          />
          <span
            style="font-size: 0.875rem;"
          >
            2023-
          </span>
        </button>
      </div>
      <div
        class="ant-divider css-dev-only-do-not-override-vrrzze ant-divider-horizontal"
        role="separator"
        style="margin: 12px 0px; background-color: rgb(118, 118, 118); height: 2px;"
      />
      <div
        class="ant-flex css-dev-only-do-not-override-vrrzze"
        style="align-items: flex-start;"
      >
        <input
          aria-label="Mark task as complete"
          class="circle-checkbox"
          style="margin-right: 12px;"
          type="checkbox"
        />
        <div
          class="ant-flex css-dev-only-do-not-override-vrrzze ant-flex-align-stretch ant-flex-justify-center ant-flex-vertical"
          style="width: 230px; height: 63px;"
        >
          <h4
            class="ant-typography css-dev-only-do-not-override-vrrzze"
            style="margin: 0px; color: rgb(255, 255, 255); font-size: 1rem; font-weight: 600; flex: 1; text-decoration: none; opacity: 1;"
          >
            Test Task
          </h4>
          <div
            class="ant-typography ant-typography-secondary css-dev-only-do-not-override-vrrzze"
            style="margin-bottom: 24px; color: rgb(255, 255, 255); font-size: 0.875rem; font-weight: 400; text-decoration: none; opacity: 1;"
          >
            Test Description
          </div>
        </div>
        <button
          aria-label="Delete task"
          class="ant-btn css-dev-only-do-not-override-vrrzze ant-btn-text ant-btn-color-default ant-btn-variant-text"
          type="button"
        >
          <img
            alt="delete icon"
            src="/src/assets/images/Delete%20Icon.png"
          />
        </button>
      </div>
      <div
        style="display: flex; justify-content: center; width: 100%; margin-top: 16px;"
      >
        <button
          aria-label="Manage task details"
          class="ant-btn css-dev-only-do-not-override-vrrzze ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
          style="width: 263px; height: 40px; font-size: 1rem; font-weight: 700; background-color: rgb(255, 255, 255); border-radius: 28px;"
          type="button"
        >
          <span>
            Manage task
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
